File: Striver/Recursion/q16.py
# print the sum of all the subsets of an array in ascending order

def subsetSum(arr, index, sum, ans):
    if index == len(arr):
        ans.append(sum)
        return
    subsetSum(arr, index+1, sum, ans)
    subsetSum(arr, index+1, sum+arr[index], ans)

arr = [1, 2, 3]
ans = []
subsetSum(arr, 0, 0, ans)
ans.sort()
print(ans)

File: Striver/Recursion/q16.java
package Striver.Recursion;

import java.util.Arrays;

public class q16 {
    // print the sum of all the subsets of an array in ascending order

    public static void main(String[] args) {
        int[] arr = { 5, 3, 2 };
        int[] ans = new int[1 << arr.length];
        int idx = 0;
        subsetSum(arr, 0, 0, ans, idx);
        Arrays.sort(ans);
        for (int i : ans) {
            System.out.print(i + " ");
        }
    }

    private static void subsetSum(int[] arr, int idx, int sum, int[] ans, int ansIdx) {
        if (idx == arr.length) {
            ans[ansIdx] = sum;
            return;
        }

        //pick
        subsetSum(arr, idx + 1, sum + arr[idx], ans, ansIdx + 1);
        
        //not pick
        subsetSum(arr, idx + 1, sum, ans, ansIdx);
    }
}
